### PR TITLE
Minimum Account Balance in Algod

### PR DESCRIPTION
**Pull request by tzaffi**
## Summary - Implementing **Minimum Balance** for `algod REST` endpoint and `goal` command line

In particular
* `/v2/accounts/{{ACCOUNT}}` has a new `min-balance` key/val
* `goal account info` returns the **Minimum Balance**
* Python SDK's `account_info` function returns `min-balance` as well

### The new  `/v2/accounts/{{ACCOUNT}}` response looks like:

```json
{
  "address": "X2MV6PXPUBPVARJO3LMTXMPN5A7IHEJTYOT6BHQ2UD3IRWRUFORDLYS4JE",
  "amount": 0,
  "amount-without-pending-rewards": 0,
  "apps-local-state": [],
  "apps-total-schema": {
    "num-byte-slice": 0,
    "num-uint": 0
  },
  "assets": [],
  "created-apps": [],
  "created-assets": [],
  "min-balance": 100000,
  "pending-rewards": 0,
  "reward-base": 0,
  "rewards": 0,
  "round": 25,
  "status": "Offline"
}
```

### The new  `goal account info` response looks like:
```bash
$ goal -w WALLET account info -a ADDRESS
Created Assets:
	<none>
Held Assets:
	<none>
Created Apps:
	ID 21, global state used 0/0 uints, 0/0 byte slices
	ID 58, global state used 0/0 uints, 0/0 byte slices
Opted In Apps:
	<none>

... etc ...
Minimum Balance:  300000 microAlgos
```

### Automated End to End Tests:

1. For `algod` and `pysdk`: [min_balance.py](./test/scripts/e2e_subs/min_balance.py) - inspired by [these testnest scenarios](https://docs.google.com/document/d/1s7_Zh0jzgokEDg-7AYJRzSRS_jMD1thKKAYOCZykgmc/edit) (but not precisely the same)
2. For `goal` [min_balance.sh](./test/scripts/e2e_subs/min_balance.sh) - a much simpler sanity check

### Followups
We need to support `min-balance` in the following SDK's:
* go-algorand-sdk
* js-algorand-sdk
* java-algorand-sdk
* Indexer should also return `min-balance`

### CODE REVIEW

**Comment by tsachiherman** [comment]
I'm not opposing to adding the min balance to the rest api endpoint, since it's a (almost) free operation. However, I'm not a big fan of the python testing - and not because of the usage of python itself. It's the circular dependency between the go-algorand and the python sdk repo that I'd like to see resolved.

File: daemon/algod/api/server/v2/handlers.go
Comment (tzaffi) [inline daemon/algod/api/server/v2/handlers.go:241]: My first time using the ledger. I'm not sure this is the ideal way of doing so to retrieve `minBalance`

File: test/scripts/e2e_subs/goal/atomic_abi.py
Comment (tzaffi) [inline test/scripts/e2e_subs/goal/atomic_abi.py:1]: WIP. Follow-up #3277 will clean this class up and make it easier to use.

File: test/scripts/e2e_subs/goal/goal.py
Comment (tzaffi) [inline test/scripts/e2e_subs/goal/goal.py:115]: @jannotti - I needed this in order to test the temporary REST endpoints while pausing my debugger.

File: test/scripts/e2e_subs/min_balance.py
Comment (tzaffi) [inline test/scripts/e2e_subs/min_balance.py:1]: Inspired by **Min Balance Test Accounts** [doc](https://docs.google.com/document/d/1s7_Zh0jzgokEDg-7AYJRzSRS_jMD1thKKAYOCZykgmc/edit)

File: test/scripts/e2e_subs/min_balance.sh
Comment (tzaffi) [inline test/scripts/e2e_subs/min_balance.sh:39]: ... yep ^^^^

File: cmd/goal/account.go
Comment (tzaffi) [inline cmd/goal/account.go:696]: Adding a line to `goal account info`'s output

File: daemon/algod/api/algod.oas2.json
Comment (tzaffi) [inline daemon/algod/api/algod.oas2.json:1370]: Suggestions for better wording greatly appreciated

File: test/scripts/e2e.sh
Comment (tzaffi) [inline test/scripts/e2e.sh:127]: I'm arguing that for testing in `go-algorand`, it's better to rely on the tip of our github than our latest official PyPi release. But honestly, in order to pull off my new `AtomicABI` Python test class, I needed this.

File: daemon/algod/api/server/v2/handlers.go
Comment (tsachiherman) [reply daemon/algod/api/server/v2/handlers.go:241]: it doesn't look that you really need the block. You only need the block header. try using `myLedger.BlockHdr`

File: test/scripts/e2e.sh
Comment (tzaffi) [inline test/scripts/e2e.sh:129]: Doesn't affect Circle tests. But of course, user beware. Make sure not to turn on the debugger when sending to Circle (...there might be a clever way to debug remotely to circle, but I don't recommend that)

File: daemon/algod/api/algod.oas2.json
Comment (tsachiherman) [reply daemon/algod/api/algod.oas2.json:1370]: The min-balance isn't just for transacting. I'd phrase it as :
```
total number of MicroAlgos required for maintaining the account
```

File: test/scripts/e2e_subs/min_balance.py
Comment (tzaffi) [inline test/scripts/e2e_subs/min_balance.py:55]: Yep, `AtomicABI` is still _alpha_. Hopefully, next week I can officially announce it (#3277 )

File: test/scripts/e2e_subs/min_balance.py
Comment (tzaffi) [inline test/scripts/e2e_subs/min_balance.py:59]: Query the `min_balance` in two ways, and compare the results and also to what we think they should be. The two ways are:
1. Via **algod REST** / **Python SDK**
2. Via **AtomicABI** to query a TEAL app that uses the `min_balance` op code

File: test/scripts/e2e_subs/tealprogs/abi-min_balance.teal
Comment (tzaffi) [inline test/scripts/e2e_subs/tealprogs/abi-min_balance.teal:2]: This TEAL ain't perfect, but it serves the purposes of this PR.

File: daemon/algod/api/algod.oas2.json
Comment (tzaffi) [reply daemon/algod/api/algod.oas2.json:1370]: 👍

File: daemon/algod/api/server/v2/handlers.go
Comment (tzaffi) [reply daemon/algod/api/server/v2/handlers.go:241]: 👍

File: test/scripts/e2e_subs/tealprogs/abi-min_balance.json
Comment (tzaffi) [inline test/scripts/e2e_subs/tealprogs/abi-min_balance.json:1]: `AtomicABI` needs the ABI contract in order to dynamically create the caller metnods.

**Comment by tzaffi** [comment]
> I'm not opposing to adding the min balance to the rest api endpoint, since it's a (almost) free operation. However, I'm not a big fan of the python testing - and not because of the usage of python itself. It's the circular dependency between the go-algorand and the python sdk repo that I'd like to see resolved.

You bring up a good point. I'm not introducing this dependency, it predates this PR. But I am certainly reinforcing it. I've also included a very simple pure shell script test for the new functionality. 

I might have stepped into a bit of a minefield here. My opinion is that the ability to write deeper tests more quickly trumps the circular dependency issue. If we ever run into the situation that PySDK makes a change that breaks our tests, we can always pin the pre-existing tests to a particular git commit and make them pass. That's a bit cumbersome, but not very hard to do.

File: daemon/algod/api/server/v2/handlers.go
Comment (algorandskiy) [inline daemon/algod/api/server/v2/handlers.go:281]: no need to calculate minBalance here. record of the basics.AccountData is passed to `AccountDataToAccount` so it can call `MinBalance` there ? I'd better pass consensus params there since it also needed for some other calculated fields if we decide to add them later.

File: test/scripts/e2e_subs/goal/atomic_abi.py
Comment (algorandskiy) [reply test/scripts/e2e_subs/goal/atomic_abi.py:1]: Is this test harness really needed for this min-balance in REST PR?

**Comment by tzaffi** [comment]
@algorandskiy - regarding your comments:
* I will investigate your and @tsachiherman suggestion about how to better acquire min balance
* Is `AtomicABI.py` _really needed_? In principle no. But to run the test in the way I wanted - compare the result of the modified `algod` with the pre-existing way we would have needed to get it from the outside (via a TEAL program), this is useful. Also, I am planning on officially introducing `AtomicABI.py` in #3277. Admittedly this is a topic of controversy. Let's continue discussing here and see if a begrudging consensus is reached.

File: daemon/algod/api/server/v2/handlers.go
Comment (tzaffi) [reply daemon/algod/api/server/v2/handlers.go:281]: Will try and incorporate your suggestions. Stay tuned.

File: daemon/algod/api/algod.oas2.json
Comment (winder) [reply daemon/algod/api/algod.oas2.json:1370]: The square brackets aren't needed for this field, the value in the brackets is the msgpack encoded name.
```suggestion
          "description": "MicroAlgo balance required by the account, the requirement starts at 1000 and grows based on the asset and application usage",
```

File: test/scripts/e2e_subs/tealprogs/abi-min_balance.teal
Comment (algonautshant) [inline test/scripts/e2e_subs/tealprogs/abi-min_balance.teal:2]: please please please add the explanation ...

File: daemon/algod/api/algod.oas3.yml
Comment (tzaffi) [reply daemon/algod/api/algod.oas3.yml:730]: this should be automatic

File: daemon/algod/api/server/v2/generated/private/types.go
Comment (tzaffi) [reply daemon/algod/api/server/v2/generated/private/types.go:52]: automatic

File: daemon/algod/api/server/v2/generated/types.go
Comment (tzaffi) [reply daemon/algod/api/server/v2/generated/types.go:52]: go generated...

File: test/scripts/e2e_subs/tealprogs/abi-min_balance.teal
Comment (tzaffi) [reply test/scripts/e2e_subs/tealprogs/abi-min_balance.teal:2]: 👍

File: test/e2e-go/cli/goal/expect/README.md
Comment (tzaffi) [inline test/e2e-go/cli/goal/expect/README.md:80]: Adding more explicit instructions for running `*.exp` tests

File: test/scripts/e2e.sh
Comment (tzaffi) [reply test/scripts/e2e.sh:129]: @jasonpaulos  @tsachiherman - pinning the python SDK version

File: test/scripts/e2e_subs/tealprogs/abi-min_balance.teal
Comment (tzaffi) [reply test/scripts/e2e_subs/tealprogs/abi-min_balance.teal]: Brief explanation added

File: daemon/algod/api/server/v2/handlers.go
Comment (algorandskiy) [inline daemon/algod/api/server/v2/handlers.go:249]: it is `internalError` rather than `notFound` since itis result does not depends on client's input

File: test/scripts/e2e.sh
Comment (algorandskiy) [reply test/scripts/e2e.sh:129]: Not sure about pinning its version, the latest and greatest looks reasonable default. Otherwise do we potentially need to update it with each pysdk release?

File: test/scripts/e2e_subs/min_balance.py
Comment (algorandskiy) [inline test/scripts/e2e_subs/min_balance.py:48]: not sure I understand why do we need all this python abi + teal machinery to test min balance field in REST

**Review by algorandskiy — COMMENTED** [review]
`MinBalance` piece in TEST looks good except `notFound` error on missed consensus.
`goal account` looks good.

> Python SDK's account_info function returns min-balance as well

I'm not against having ABI helpers for e2e tests. I only do not understand why do we test sdk in go-algorand. I think it should be another way around

File: daemon/algod/api/server/v2/handlers.go
Comment (tzaffi) [reply daemon/algod/api/server/v2/handlers.go:249]: 👍

**Comment by tzaffi** [comment]
> `MinBalance` piece in TEST looks good except `notFound` error on missed consensus. `goal account` looks good.
> 
> > Python SDK's account_info function returns min-balance as well
> 
> I'm not against having ABI helpers for e2e tests. I only do not understand why do we test sdk in go-algorand. I think it should be another way around

I did not intend to test the Python SDK in this PR. However, that comes for free because I'm using the Python Goal wrapper which internally calls the Python SDK. So I am pointing this fact out so that no extra work needs to be done for the Python SDK.

You make a good point that we shouldn't be testing SDK's in go-algorand. So there will need to be appropriate tests added for **min-balance** in `algorand-sdk-testing` and each of the SDK's will need to pass those tests. The fact that the Python SDK should be good to go, doesn't excuse it from passing those tests.

I will be addressing your point regarding the error type shortly.

**Review by algorandskiy — DISMISSED** [review]
The `MinBalance` piece looks good.
python sdk pinning and testing harness - defer to others.

File: daemon/algod/api/algod.oas2.json
Comment (tsachiherman) [inline daemon/algod/api/algod.oas2.json:1370]: I think it would be better to drop off the numeric constants from the description. These values are consensus protocol dependent and can be modified in the future.

File: daemon/algod/api/server/v2/account_test.go
Comment (tsachiherman) [inline daemon/algod/api/server/v2/account_test.go:105]: The minimum balance depends of a multiple fields from `config.ConsensusParams`. It would be wrong to initialize just a single field ( `MinBalance` ), since the function `record.MinBalance` might change over time.
I would suggest you'll grab a specific consensus protocol and use it. i.e. the current protocol sounds like a safe choice here.

File: daemon/algod/api/server/v2/account_test.go
Comment (tsachiherman) [inline daemon/algod/api/server/v2/account_test.go:199]: same as above.

File: test/scripts/e2e_subs/min_balance.py
Comment (tsachiherman) [reply test/scripts/e2e_subs/min_balance.py:48]: I agree with @algorandskiy about that one. This python testing should go to the python repository.
Testing the REST endpoint of the algod can easily be done using either expect or one of the existing e2e subs.
( you've already implemented min_balance.sh, which is great )

File: test/scripts/e2e_subs/tealprogs/abi-min_balance.json
Comment (tsachiherman) [inline test/scripts/e2e_subs/tealprogs/abi-min_balance.json:30]: nit: add EOL

File: test/scripts/e2e_subs/tealprogs/abi-min_balance.json
Comment (ahangsu) [inline test/scripts/e2e_subs/tealprogs/abi-min_balance.json:3]: nit:  😅 we are changing this field `"appId"` to `network` in the new ABI spec: https://github.com/algorandfoundation/ARCs/pull/53, it seems totally optional to have this field or not, might be okay to remove that

File: test/scripts/e2e_subs/tealprogs/abi-min_balance.json
Comment (ahangsu) [inline test/scripts/e2e_subs/tealprogs/abi-min_balance.json:28]: nit: seems not essential in this test, can remove in contract file and teal code

File: daemon/algod/api/server/v2/account_test.go
Comment (tzaffi) [reply daemon/algod/api/server/v2/account_test.go:105]: 👍

File: daemon/algod/api/algod.oas2.json
Comment (tzaffi) [reply daemon/algod/api/algod.oas2.json:1370]: 👍

File: daemon/algod/api/server/v2/account_test.go
Comment (tzaffi) [reply daemon/algod/api/server/v2/account_test.go:199]: 👍

File: test/scripts/e2e_subs/tealprogs/abi-min_balance.json
Comment (tzaffi) [reply test/scripts/e2e_subs/tealprogs/abi-min_balance.json:30]: 👍

File: test/scripts/e2e_subs/tealprogs/abi-min_balance.json
Comment (tzaffi) [reply test/scripts/e2e_subs/tealprogs/abi-min_balance.json:3]: 👍

File: test/scripts/e2e_subs/tealprogs/abi-min_balance.json
Comment (tzaffi) [reply test/scripts/e2e_subs/tealprogs/abi-min_balance.json:30]: 👍

File: daemon/algod/api/server/v2/account_test.go
Comment (tzaffi) [reply daemon/algod/api/server/v2/account_test.go:199]: 👍

File: test/scripts/e2e_subs/tealprogs/abi-min_balance.json
Comment (tzaffi) [reply test/scripts/e2e_subs/tealprogs/abi-min_balance.json:28]: 👍

File: test/scripts/e2e_subs/goal/goal.py
Comment (jannotti) [reply test/scripts/e2e_subs/goal/goal.py:115]: I don't quite get it.  The info was already available as `goal.algod.algod_address` now it's reformatted so you can say `val = goal.get_endpoint_info(); val["algod"]["url"]` ?

File: test/scripts/e2e_subs/min_balance.py
Comment (jannotti) [reply test/scripts/e2e_subs/min_balance.py:60]: I'm in favor of using python instead of an unholy combination or shell/goal/jq/curl/grep etc, so tests are clearer.  But I think the ABI stuff is overboard here. A very, very simple teal app that returns min_balance could be instantiated and called using just the sdk basics that were already here.

As a general rule, if we just use python as a more convenient way to write good tests, and stay away from testing the most recent features (ABI in this case) we will get better tests, but we can avoid difficulties with pinning the version.  Just use old tried and true SDK functions.

File: test/scripts/e2e_subs/min_balance.py
Comment (tzaffi) [reply test/scripts/e2e_subs/min_balance.py]: @algonautshant and @tsachiherman  - please have a look again. I've implemented the decision points from yesterday's meeting:

1. Remove dependence on Python SDK's ABI interaction implementation, and in particular, removed the `AtomicABI` test class
2. Keeping the python tests, but refactored to avoid ABI
3. Pinned the Python SDK and added a comment of explanation

**Review by algorandskiy — COMMENTED** [review]
LGTM, please implement @tsachiherman 's suggestions

File: test/scripts/e2e_subs/goal/goal.py
Comment (tzaffi) [reply test/scripts/e2e_subs/goal/goal.py:115]: 👍

Comment (tzaffi) [reply test/scripts/e2e_subs/goal/goal.py:115]: Just making it a bit more straightforward to get all the info, but I'll revert.

File: test/scripts/e2e_subs/min_balance.py
Comment (tzaffi) [reply test/scripts/e2e_subs/min_balance.py]: 👍

File: test/scripts/e2e_subs/min_balance.py
Comment (tzaffi) [inline test/scripts/e2e_subs/min_balance.py:44]: @jannotti - simpler TEAL program not using ABI

File: daemon/algod/api/server/v2/account_test.go
Comment (tzaffi) [inline daemon/algod/api/server/v2/account_test.go:105]: @tsachiherman : `proto` is **ConsensusFuture**. That seems like a reasonable option.

**Review by algorandskiy — DISMISSED** [review]

File: test/e2e-go/cli/goal/expect/README.md
Comment (tsachiherman) [reply test/e2e-go/cli/goal/expect/README.md:80]: interesting.. I was always using the following syntax:
```sh
cd test/e2e-go/cli/goal/expect
go test -v . --run TestGoalWithExpect/goalCmdFlagsTest.exp
```

**Review by tsachiherman — DISMISSED** [review]

**Review by barnjamin — DISMISSED** [review]

File: daemon/algod/api/server/v2/handlers.go
Comment (jannotti) [inline daemon/algod/api/server/v2/handlers.go:250]: Nicer to use `func (l *Ledger) ConsensusParams(r basics.Round) (config.ConsensusParams, error)`

File: test/scripts/e2e_subs/min_balance.py
Comment (jannotti) [reply test/scripts/e2e_subs/min_balance.py:44]: I'd use version 5 here.  We're probably going to release this before teal 6 is available.  Nothing will break, since this is only a test, but it seems weird to test it with a v6 teal script even though it's a feature that comes out before v6.

File: daemon/algod/api/server/v2/handlers.go
Comment (tzaffi) [reply daemon/algod/api/server/v2/handlers.go:250]: 👍

File: test/scripts/e2e_subs/min_balance.py
Comment (tzaffi) [inline test/scripts/e2e_subs/min_balance.py:143]:
```suggestion
expected_goal_mb = CONSENSUS_MIN_BALANCE + APP_MIN_BALANCE

# starting out, should be at global min
```

File: test/scripts/e2e_subs/min_balance.py
Comment (tzaffi) [inline test/scripts/e2e_subs/min_balance.py:141]:
```suggestion
```

**Review by jannotti — APPROVED** [review]
Looks good.  I'll wait for the CI checks to finish, then merge.

File: test/e2e-go/cli/goal/expect/README.md
Comment (tzaffi) [reply test/e2e-go/cli/goal/expect/README.md:80]: ohhh... I only saw this now. Will try to get this into the README next PR.

**Pull request merged**
